sigstore/dsse/_predicate.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2024 The Sigstore Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Models for the predicates used in in-toto statements
+"""
+
+from typing import Any, Dict, List, Literal, Optional, TypeVar, Union
+
+from pydantic import (
+    BaseModel,
+    Field,
+    RootModel,
+    StrictBytes,
+    StrictStr,
+    model_validator,
+)
+
+from sigstore.dsse import Digest
+
+PREDICATE_TYPE_SLSA_v0_2 = "https://slsa.dev/provenance/v0.2"
+PREDICATE_TYPE_SLSA_v1_0 = "https://slsa.dev/provenance/v1"
+
+SUPPORTED_PREDICATE_TYPES = [PREDICATE_TYPE_SLSA_v0_2, PREDICATE_TYPE_SLSA_v1_0]
+
+# Common models
+
+DigestSetSource = RootModel[Dict[Union[Digest, Literal["sha1"]], str]]
+"""
+Same as `dsse.DigestSet` but with `sha1` added.
+
+Since this model is not used to verify hashes, but to parse predicates that might
+contain hashes, we include this weak hash algorithm. This is because provenance
+providers like GitHub use SHA1 in their predicates to refer to git commit hashes.
+"""
+
+
+class Predicate(BaseModel):
+    """
+    Base model for in-toto predicates
+    """
+
+    pass
+
+
+# Models for SLSA Provenance v0.2
+
+
+class BuilderV0_1(BaseModel):
+    """
+    The Builder object used by SLSAPredicateV0_2
+    """
+
+    id: StrictStr
+
+
+class ConfigSource(BaseModel):
+    """
+    The ConfigSource object used by Invocation in v0.2
+    """
+
+    uri: Optional[StrictStr]
+    digest: Optional[DigestSetSource]
+    entry_point: Optional[StrictStr] = Field(None, alias="entryPoint")
```

Comment:
Nitpick: Rather than having to explicitly declare each `alias` here and below, maybe we can define a common base model with a `model_config` that tweaks alias generation?

Something like:

```python
from pydantic.config import ConfigDict
from pydantic.alias_generators import to_camel

class _PredicateBase(BaseModel):
    model_config = ConfigDict(alias_generator=to_camel)

class ConfigSource(_PredicateBase):
    ...
```

...I think something like that will work, as will save us some manual typing 🙂

Reply:
Good idea! I also fixed a couple of bugs I found:
- `slsa-framework/slsa-github-generator` incorrectly generates predicates using `buildInvocationID` (note the final uppercase D, should be `buildInvocationId` instead). In order to fix that, I added a validation alias that allows for both options. The serialization alias will use the camelCase one set by the config you suggested.
- I was dumping the model without specifying it should output the camelCase aliases rather than our snake_case field names. Fixed by passing `by_alias=True` to `model_dump()`
- Our model dump also included the optional fields that were set to `None`. However, that's how we model them, those fields are just absent in the source predicate. So now we pass `exclude_none=True` to `model_dump`, to exclude those fields from the dump.

Reply:
Mmm given the last point, I wonder if instead of dumping our Predicate model, we should instead just validate it and then use the exact input the user gave us to build the in-toto statement. 
Otherwise we might end up modifying it when doing the JSON -> Model -> JSON roundtrip.

Reply:
@woodruffw WDYT about this?

Reply:
@facutuesca I think that's a good idea! In principle the each JSON layout shouldn't need to be stable, but in practice users might expect that (and as you've pointed out it avoids a potential source of ser/de errors on our side).